demos/rpmsg_multi_services.rst

```diff
@@ -0,0 +1,154 @@
+
+The main target of the demonstration is to show the different RPMsg types as supported by Linux drivers, namely 'direct', 'character' and 'tty' driver.
+
+The underlying OpenAMP architectural components used by these applications are
+
+* :ref:`Remoteproc<overview-remoteproc-work-label>`
+* :ref:`Resource Table<overview-remoteproc-work-label>`
+* :ref:`RPMsg<overview-rpmsg-work-label>`
+* :ref:`Virtio<overview-rpmsg-work-label>`
+* :ref:`Libmetal<overview-proxy-libmetal-label>`
+
+The supporting Linux architectural components used by the drivers on the host side are
+
+* `RPMsg character device <https://linux-kernel-labs.github.io/refs/heads/master/labs/device_drivers.html>`_
+* `tty device <https://www.kernel.org/doc/html/latest/driver-api/tty/index.html>`_
+
+The following architecture diagram shows the components involved in the demonstration.
+
+..  image::  ../images/demos/rpmsg-multi-services-components.svg
+
+.. _rpmsg-control-flow-label:
+
+The top-level control flow is shown in the following message diagram. The remote threads are show sequentially for clarity of diagram but could be executed in parallel.
+
+..  image::  ../images/demos/rpmsg-multi-services-control-flow.svg
+
+
+.. _rpmsg-client-sample-label:
+
+RPmsg Client Sample
+===================
+
+The Linux rpmsg_client_sample driver begins sending 'hello world!' messages on a rpmsg_driver probe, initiated by a name service announcement from the remote. This is repeated a predefined count times for each response from the remote. The response from the remote application is to return the same packet received at the :ref:`RPmsg endpoint <rpmsg-endpoint>` of the host controller.
+
+When the count (100) responses have been sent, the endpoint is destroyed by the remote.
+
+
+.. _rpmsg-character-driver-sample-label:
+
+Character Driver Samples
+========================
+
+The Linux character and tty driver are demonstrated by sending messages to the rpmsg_char and rpmsg_tty driver respectively by a user space command writing to the respective /dev/rpmsg? and /dev/ttyRPMSG? device files. The response is output by using the Linux cat command.
+
+Writing data to the zeroth index /dev/rpmsg0 and /dev/ttyRPMSG0 files will result in an RPMsg being sent to the channel's address, which start at 0x400 (1024) and are allocated by the remote sequentially on channel creation.
+
+Writing data to indexed driver files will result in an RPMsg being sent the Linux endpoint address, which has a corresponding address on the remote side in either a one to one or many to one relationship as detailed in the following sections.
+
+.. _rpmsg-raw-driver-label:
+
+Raw Character Driver Sample
+---------------------------
+
+When started the character/raw remote service (app_rpmsg_raw thread) creates two RPMsg endpoints. The first with the special RPMSG_ADDR_ANY (-1) address which sets up the RPMsg channel and the second with destination and source address set to 1.
```

Comment:
Use comma to separate introductory element "When started" from the main part of the sentence. TODO: check for other occurrences.